stdlib/public/core/StringWordView.swift

```diff
@@ -0,0 +1,202 @@
+//
+// This source file is part of the Swift.org open source project
+//
+// Copyright (c) 2022 Apple Inc. and the Swift project authors
+// Licensed under Apache License v2.0 with Runtime Library Exception
+//
+// See https://swift.org/LICENSE.txt for license information
+//
+//===----------------------------------------------------------------------===//
+
+extension String {
+  @_spi(_Unicode)
+  @available(SwiftStdlib 5.7, *)
+  public struct _WordView {
+    internal var _guts: _StringGuts
+
+    internal init(_ _guts: _StringGuts) {
+      self._guts = _guts
+    }
+  }
+}
+
+@available(SwiftStdlib 5.7, *)
+extension String._WordView: Collection {
+  @_spi(_Unicode)
+  @available(SwiftStdlib 5.7, *)
+  public typealias Index = String.Index
+
+  @_spi(_Unicode)
+  @available(SwiftStdlib 5.7, *)
+  public typealias Element = Substring
+
+  @_spi(_Unicode)
+  @available(SwiftStdlib 5.7, *)
+  public var startIndex: Index {
+    _guts.startIndex
+  }
+
+  @_spi(_Unicode)
+  @available(SwiftStdlib 5.7, *)
+  public var endIndex: Index {
+    _guts.endIndex
+  }
+
+  @_spi(_Unicode)
+  @available(SwiftStdlib 5.7, *)
+  public func index(after i: Index) -> Index {
+    let i = _guts.validateWordIndex(i)
+    return _uncheckedIndex(after: i)
+  }
+
+  internal func _uncheckedIndex(after i: Index) -> Index {
+    _internalInvariant(_guts.hasMatchingEncoding(i))
+    _internalInvariant(i < endIndex)
+    _internalInvariant(i._isWordAligned)
+
+    if _slowPath(_guts.isForeign) {
+      return _foreignIndex(after: i)
+    }
+
+    let nextOffset = _guts.withFastUTF8 { utf8 in
+      nextBoundary(startingAt: i._encodedOffset) {
+        _internalInvariant($0 >= 0)
+
+        guard $0 < utf8.count else {
+          return nil
+        }
+
+        let (scalar, len) = _decodeScalar(utf8, startingAt: $0)
+        return (scalar, $0 &+ len)
+      }
+    }
+
+    let nextIndex = String.Index(_encodedOffset: nextOffset)._wordAligned
+    return _guts.markEncoding(nextIndex)
+  }
+
+  @inline(never)
+  internal func _foreignIndex(after i: Index) -> Index {
+#if _runtime(_ObjC)
+    let nextOffset = nextBoundary(startingAt: i._encodedOffset) {
+      _internalInvariant($0 >= 0)
+
+      guard $0 < _guts.count else {
+        return nil
+      }
+
+      let scalars = String.UnicodeScalarView(_guts)
+      let idx = String.Index(_encodedOffset: $0)
+
+      let scalar = scalars[idx]
+      let nextIndex = scalars.index(after: idx)
+
+      return (scalar, nextIndex._encodedOffset)
+    }
+
+    let nextIndex = String.Index(_encodedOffset: nextOffset)._scalarAligned
+    return _guts.markEncoding(nextIndex)
+#else
+    fatalError("No foreign strings on this platform in this version of Swift.")
+#endif
+  }
+
+  @_spi(_Unicode)
+  @available(SwiftStdlib 5.7, *)
+  public subscript(position: Index) -> Element {
+    let position = _guts.validateWordIndex(position)
+    let indexAfter = _uncheckedIndex(after: position)
```

Comment:
It would be useful to know if caching the word size in the index (like we do in the character views) would have a performance benefit here, and if so, how much it matters.

Reply:
Usually more important to implement `Iterator`. But this index could store a range for the word, if that makes sense. (Haven't thought through it).